### PR TITLE
Zen2: Cluster state publication pipeline

### PR DESCRIPTION
**Pull request by ywelsch**
This PR implements the state machine on the master to publish a cluster state.

Relates to #32006

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    private void initializeCluster(VotingConfiguration initialConfig) {
+        node1.coordinationState.setInitialState(CoordinationStateTests.clusterState(0L, 1L, n1, initialConfig, initialConfig, 0L));
+        StartJoinRequest startJoinRequest = new StartJoinRequest(n1, 1L);
+        node1.coordinationState.handleJoin(node1.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node2.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node3.coordinationState.handleStartJoin(startJoinRequest));
+        assertTrue(node1.coordinationState.electionWon());
+    }
+
+    public void testSimpleClusterStatePublishing() throws InterruptedException {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Could this make finer assertions about the state that the publication is in? More precise requests follow...

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        DiscoveryNodes discoveryNodes = DiscoveryNodes.builder().add(n1).add(n2).add(n3).localNodeId(n1.getId()).build();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, Collections.emptySet());
+
+        assertThat(publication.pendingPublications.keySet(), equalTo(discoNodes));
+        assertTrue(publication.pendingCommits.isEmpty());
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                publication.publishRequest);
+            e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Could this assert that `publication.applyCommit` is null iff `n1` hasn't been processed?

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+        });
+
+        assertThat(publication.pendingCommits.keySet(), equalTo(discoNodes));
+        assertNotNull(publication.applyCommit);
+        assertEquals(publication.applyCommit.getTerm(), publication.publishRequest.getAcceptedState().term());
+        assertEquals(publication.applyCommit.getVersion(), publication.publishRequest.getAcceptedState().version());
+        publication.pendingCommits.entrySet().stream().collect(shuffle()).forEach(e -> {
+            nodeResolver.apply(e.getKey()).coordinationState.handleCommit(publication.applyCommit);
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Could we assert `publication.completed == false` and `publication.committed == false` at the top of this lambda?

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+            e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+        });
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        assertThat(ackListener.await(0L, TimeUnit.SECONDS).size(), equalTo(3));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Could this be:

    assertThat(ackListener.await(0L, TimeUnit.SECONDS), containsInAnyOrder(n1, n2, n3));

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        node1.coordinationState.setInitialState(CoordinationStateTests.clusterState(0L, 1L, n1, initialConfig, initialConfig, 0L));
+        StartJoinRequest startJoinRequest = new StartJoinRequest(n1, 1L);
+        node1.coordinationState.handleJoin(node1.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node2.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node3.coordinationState.handleStartJoin(startJoinRequest));
+        assertTrue(node1.coordinationState.electionWon());
+    }
+
+    public void testSimpleClusterStatePublishing() throws InterruptedException {
+        VotingConfiguration singleNodeConfig = new VotingConfiguration(Sets.newHashSet(n1.getId()));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Could we also have a demonstration of the happy path on a three-node configuration, with the assertions adjusted accordingly? In the three-node case it's possible that publication responses interleave with commits, and this should be covered.

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+            failNodeWhenCommitting, publicationDidNotMakeItToNode2);
+
+        Set<DiscoveryNode> initialFaultyNodes =
+            failNodeWhenCommitting == false && remainingActions.decrementAndGet() == injectFaultAt ?
+                Collections.singleton(n2) : Collections.emptySet();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, initialFaultyNodes);
+
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (failNodeWhenCommitting == false) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: We could get rid of this `== false` by inverting this statement.

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+                if (e.getKey().equals(n2) == false || publicationDidNotMakeItToNode2 == false) {
+                    PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                        publication.publishRequest);
+                    e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+                }
+            }
+        });
+
+        publication.pendingCommits.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (failNodeWhenCommitting == false) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: We could get rid of this `== false` by inverting this statement.

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+            e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+        });
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        assertThat(ackListener.await(0L, TimeUnit.SECONDS).size(), equalTo(3));
+    }
+
+    public void testClusterStatePublishingWithFaultyNode() throws InterruptedException {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: I think that this test is trying to test too many things via its randomly-set flags, making it hard to tell what it's actually covering. Please could we split out the different cases into separate tests a bit?

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        }
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        publication.onFaultyNode(randomFrom(n1, n3)); // has no influence
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(1));
+        assertThat(errors.get(0).v1(), equalTo(n2));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Can we also assert something about the error, and also assert that the successful acks are as expected.

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+            }
+        });
+
+        assertThat(publication.pendingCommits.keySet(), equalTo(Collections.emptySet()));
+        assertNull(publication.applyCommit);
+        assertTrue(publication.completed);
+        assertFalse(publication.committed);
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(3));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Can we assert something about the errors themselves, and that there were no successful acks? (also throughout the other tests)

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class PublicationTests extends ESTestCase {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: As a general comment, I wonder if using the `DeterministicTaskQueue` for these tests would make them look a bit simpler while allow for more interleaving than we currently do. Perhaps a followup like with the ConsensusState testing?

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+
+        @Override
+        protected void onCompletion(boolean committed) {
+            completed = true;
+            this.committed = committed;
+        }
+
+        @Override
+        protected void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response) {
+
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: We should assert that this is called (and I guess that we can say something about the response that is passed in).

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+        Map<DiscoveryNode, ActionListener<TransportResponse.Empty>> pendingCommits = new HashMap<>();
+
+        MockPublication(Settings settings, PublishRequest publishRequest, Discovery.AckListener ackListener,
+                               LongSupplier currentTimeSupplier) {
+            super(settings, publishRequest, ackListener, currentTimeSupplier);
+            this.publishRequest = publishRequest;
+        }
+
+        @Override
+        protected void onCompletion(boolean committed) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Can we assert that this is only ever called once?

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+            isCompleted = true;
+            onCompletion(false);
+            return;
+        }
+
+        assert isCompleted == false;
+        isCompleted = true;
+        onCompletion(true);
+        assert applyCommitRequest.isPresent();
+        logger.trace("onPossibleCompletion: [{}] was successful, applying new state locally", this);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: not sure it makes sense to talk about `applying` the state here

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+        }
+    }
+
+    protected abstract void onCompletion(boolean committed);
+
+    protected abstract boolean isPublishQuorum(CoordinationState.VoteCollection votes);
+
+    protected abstract Optional<ApplyCommitRequest> handlePublishResponse(DiscoveryNode sourceNode, PublishResponse publishResponse);
+
+    protected abstract void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: Does this need the whole `PublishWithJoinResponse` or could it just take the `Optional<Join>`? Or maybe even the `Join` only if present?

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+
+        @Override
+        public String toString() {
+            return state.toString();
+        }
+    }
+
+    private class PublicationTarget {
+        private final DiscoveryNode discoveryNode;
+        private final PublicationTargetStateMachine publicationTargetStateMachine = new PublicationTargetStateMachine();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: This was useful for development, but now things are more stable shall we merge this into `PublicationTarget`?

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+            return discoveryNode.getId();
+        }
+
+        public void sendPublishRequest() {
+            if (publicationTargetStateMachine.isFailed()) {
+                return;
+            }
+            publicationTargetStateMachine.setState(PublicationTargetState.SENT_PUBLISH_REQUEST);
+            Publication.this.sendPublishRequest(discoveryNode, publishRequest, new PublicationTarget.PublishResponseHandler());
+            // TODO Can this ^ fail with an exception? Target should be failed if so.
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: Does this question have an answer? I _think_ any failures are passed to the response handler.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I mostly asked for changes around the tests, but there's a couple of things in `Publication` too.

There's a handful of unnecessarily-`public` methods according to IntelliJ that I think can be more restricted. I haven't looked at the privacy levels in detail yet.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java (hunk 17 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: I'm deferring this TODO until we integrate this with Legislator.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java (hunk 15 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: I don't know yet. I've added a TODO in 43e3399f164

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java (hunk 16 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: ok, fixed in b7b4e39b980

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java (hunk 14 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: fixed in 25d5835ea8c

File: server/src/main/java/org/elasticsearch/cluster/coordination/PublishWithJoinResponse.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * Response to a {@link PublishRequest}. Encapsulates both a {@link PublishResponse}
+ * and an optional {@link Join}.
+ */
+public class PublishWithJoinResponse extends TransportResponse {
+    private final PublishResponse publishResponse;
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/PublishWithJoinResponse.java]: I've removed and inlined TermVersionResponse, see a21e8c7b849

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 13 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: sure, done in f09659ceda6

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 12 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: sure, done in f09659ceda6

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 2 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: sure, fixed in f09659ceda6

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 3 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: sure, fixed in f09659ceda6

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 4 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: jup (f09659ceda6)

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 8 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: I've split this into two tests in a76c4f1c7cd

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 11 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: I wonder if that part is better covered by LegislatorTests, in particular as it hooks in more interesting ConsensusState instances

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 10 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: sure, gave this a try in b711edb4894

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 6 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: this became obsolete by splitting test into two

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 7 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: this became obsolete by splitting test into two

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 5 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Even in a single-node configuration cluster, publish responses can arrive after commit responses. We already test here that publish responses arrive before committing and after committing. We don't test commit responses arriving before all publish responses are handled. I've added this in  0199b3de7b2

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 1 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: see comments below

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+        SENT_APPLY_COMMIT,
+        APPLIED_COMMIT,
+    }
+
+    private class PublicationTarget {
+        private final DiscoveryNode discoveryNode;
+        private boolean ackIsPending = true;
+        private PublicationTargetState state = PublicationTargetState.NOT_STARTED;
+
+        private PublicationTarget(DiscoveryNode discoveryNode) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: Not 100% sure of the rules here, but `private` seems too restricted for this.

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 20]
```diff
@@ -0,0 +1,429 @@
...
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(3));
+        assertThat(errors.stream().map(Tuple::v1).collect(Collectors.toList()), containsInAnyOrder(n1, n2, n3));
+        errors.stream().forEach(tuple ->
+            assertThat(tuple.v2().getMessage(), containsString(timeOut ? "timed out" :
+                tuple.v1().equals(n2) ? "dummy failure" : "non-failed nodes do not form a quorum")));
+    }
+
+    public void testClusterStatePublishingTimesOutAfterCommit() throws InterruptedException {
+        VotingConfiguration config = new VotingConfiguration(Sets.newHashSet(n1.getId(), n2.getId()));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: Could we (randomly?) add `n3` to the configuration here?

**Review by DaveCTurner — COMMENTED** [review]
I find it a little strange that we don't have _any_ tests with more than 2 nodes in the configuration, to demonstrate that we only need a majority of responses to commit. I don't think this is a gap in coverage so much as in documentation. I suggested a fix for this, and also found one `private` thing that I think shouldn't be.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java (hunk 19 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java]: fixed by 799157d72d4

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java (hunk 20 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: sure, fixed in 799157d72d4

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by ywelsch** [comment]
Thanks @DaveCTurner

**Pull request merged**